File: packages/python/procedurals/noise_band_path.py
"""Attempt-bounded scalar-band walk independently specified by the shared
path.noise-band-trace-2d contract.

Motivating note: survey/out/2018/Generativos/venas/notes.md. No artistic
attempts/tolerance/scale range is established; see the catalog contract for
evidence. Uses a bit-exact ported fdlibm5.3 sin/cos (verified against
java.lang.StrictMath.sin/cos), not host math.sin/cos.
"""
from array import array
from dataclasses import dataclass
import copy
import math

from .fields import gradient_noise_2d_01
from ._fdlibm_trig import fdlibm_cos, fdlibm_sin

__all__ = ["NoiseBandPathError", "noise_band_path_2d"]
_HALF_PI = math.pi / 2  # exactly 0x1.921fb54442d18p0 (halving is exact)
_UINT32_MAX = 4294967295
_INT32_MAX = 2147483647
_MAX_VERTICES = 1073741823
_MAX_SAFE = 9007199254740991
_KEYS = {"field", "start", "heading", "seed", "attempts", "stepDistance",
          "fieldScale", "fieldOffset", "tolerance", "maxVertices"}
_MASK32 = (1 << 32) - 1
_MASK64 = (1 << 64) - 1


class NoiseBandPathError(ValueError):
    """Stable catalog code, with attemptIndex/stage for the three dynamic codes."""
    def __init__(self, code, attempt_index=None, stage=None):
        super().__init__(code)
        self.code = code
        if code in ("TRACE_ARITHMETIC_INVALID", "TRACE_QUERY_INVALID", "VERTEX_LIMIT_EXCEEDED"):
            self.attempt_index = attempt_index
            self.stage = stage


def _fail(code):
    raise NoiseBandPathError(code)


def _zero(value):
    return 0.0 if value == 0 else value


def _num(value):
    if type(value) not in (int, float) or type(value) is bool:
        _fail("INVALID_INPUT")
    try:
        value = float(value)
    except OverflowError:
        raise NoiseBandPathError("INVALID_INPUT") from None
    if not math.isfinite(value):
        _fail("INVALID_INPUT")
    return value


def _uint(value):
    n = _num(value)
    if n < 0 or n > _UINT32_MAX or n != math.floor(n):
        _fail("INVALID_INPUT")
    return int(n)


def _integer(value, lo, hi):
    n = _num(value)
    if n < lo or n > hi or n != math.floor(n):
        _fail("INVALID_INPUT")
    return int(n)


def _nonnegative(value):
    n = _num(value)
    if n < 0:
        _fail("INVALID_INPUT")
    return n


def _pair(value):
    if type(value) is not list or len(value) != 2:
        _fail("INVALID_INPUT")
    return _zero(_num(value[0])), _zero(_num(value[1]))


def _arith(value, attempt_index, stage):
    if not math.isfinite(value):
        raise NoiseBandPathError("TRACE_ARITHMETIC_INVALID", attempt_index, stage)
    return value


def _query(field, x, y, scale, offset_x, offset_y, attempt_index, prefix):
    """Q(x,y,prefix): field query with strict per-axis multiply-then-add domain checks."""
    qx = x * scale
    if not math.isfinite(qx):
        raise NoiseBandPathError("TRACE_QUERY_INVALID", attempt_index, prefix + "_x")
    qx = qx + offset_x
    if not math.isfinite(qx) or qx < -_MAX_SAFE or qx >= _MAX_SAFE:
        raise NoiseBandPathError("TRACE_QUERY_INVALID", attempt_index, prefix + "_x")
    qy = y * scale
    if not math.isfinite(qy):
        raise NoiseBandPathError("TRACE_QUERY_INVALID", attempt_index, prefix + "_y")
    qy = qy + offset_y
    if not math.isfinite(qy) or qy < -_MAX_SAFE or qy >= _MAX_SAFE:
        raise NoiseBandPathError("TRACE_QUERY_INVALID", attempt_index, prefix + "_y")
    return field.sample(qx, qy)


class _Xoshiro:
    """Private xoshiro128**1.1 stream with two SplitMix64 expansion words;
    identical algorithm to procedurals.quadrant_partition._Stream."""
    def __init__(self, seed):
        x = seed + 0x9E3779B97F4A7C15
        a = self._split(x)
        b = self._split(x + 0x9E3779B97F4A7C15)
        self._s = [a & _MASK32, a >> 32, b & _MASK32, b >> 32]

    @staticmethod
    def _split(x):
        x &= _MASK64
        x = ((x ^ (x >> 30)) * 0xBF58476D1CE4E5B9) & _MASK64
        x = ((x ^ (x >> 27)) * 0x94D049BB133111EB) & _MASK64
        return (x ^ (x >> 31)) & _MASK64

    def output(self):
        s = self._s
        value = (s[1] * 5) & _MASK32
        result = ((((value << 7) | (value >> 25)) & _MASK32) * 9) & _MASK32
        t = (s[1] << 9) & _MASK32
        s[2] ^= s[0]
        s[3] ^= s[1]
        s[1] ^= s[2]
        s[0] ^= s[3]
        s[2] ^= t
        s[3] = ((s[3] << 11) | (s[3] >> 21)) & _MASK32
        return result

    def unit(self):
        return self.output() / 4294967296.0


@dataclass(frozen=True, slots=True)
class _Path:
    _positions: array
    _headings: array
    _attempts: int
    _accepted: int
    _rejected: int
    _config: dict

    @property
    def size(self):
        return len(self._positions) // 2

    @property
    def attempts(self):
        return self._attempts

    @property
    def accepted(self):
        return self._accepted

    @property
    def rejected(self):
        return self._rejected

    def _checked(self, index, headings):
        if type(index) not in (int, float) or type(index) is bool:
            _fail("INVALID_INDEX")
        if type(index) is float and (not math.isfinite(index) or not index.is_integer()):
            _fail("INVALID_INDEX")
        if index < 0 or index > _MAX_SAFE:
            _fail("INVALID_INDEX")
        index = int(index)
        if index >= (self._accepted if headings else self.size):
            _fail("INDEX_OUT_OF_RANGE")
        return index

    def point_at(self, index):
        """Return a fresh detached [x, y] binary64 point."""
        i = self._checked(index, False)
        return [self._positions[2 * i], self._positions[2 * i + 1]]

    def point_into(self, index, output, offset=0):
        """Validate index and both destination slots before writing; return output."""
        i = self._checked(index, False)
        if not (type(output) is list or (type(output) is array and output.typecode == "d" and output.itemsize == 8)):
            _fail("INVALID_OUTPUT")
        if type(offset) not in (int, float) or type(offset) is bool:
            _fail("INVALID_OUTPUT")
        if type(offset) is float and (not math.isfinite(offset) or not offset.is_integer()):
            _fail("INVALID_OUTPUT")
        offset = int(offset)
        if offset < 0 or offset > len(output) - 2:
            _fail("INVALID_OUTPUT")
        output[offset] = self._positions[2 * i]
        output[offset + 1] = self._positions[2 * i + 1]
        return output

    def heading_at(self, index):
        """Return the accepted-segment proposal heading."""
        i = self._checked(index, True)
        return self._headings[i]

    def serialize(self):
        """Return a detached configuration snapshot for replay."""
        return copy.deepcopy(self._config)

    def to_values(self):
        """Export detached positions, headings and counters; retain no export objects."""
        return {
            "positions": [self.point_at(i) for i in range(self.size)],
            "headings": list(self._headings),
            "attempts": self._attempts,
            "accepted": self._accepted,
            "rejected": self._rejected,
        }


def noise_band_path_2d(config):
    """Retain an attempt-bounded connected path whose accepted proposals remain
    within a strict scalar band around the starting noise value.

    Implements path.noise-band-trace-2d 0.1.0 independently of source code.
    """
    if type(config) is not dict or set(config) != _KEYS:
        _fail("INVALID_INPUT")
    field_config = config["field"]
    if type(field_config) is not dict or set(field_config) != {"seed"}:
        _fail("INVALID_INPUT")
    field_seed = _uint(field_config["seed"])
    start_x0, start_y0 = _pair(config["start"])
    heading = _zero(_num(config["heading"]))
    seed = _uint(config["seed"])
    tries = _integer(config["attempts"], 0, _INT32_MAX)
    distance = _zero(_nonnegative(config["stepDistance"]))
    scale = _zero(_num(config["fieldScale"]))
    offset_x, offset_y = _pair(config["fieldOffset"])
    tolerance = _zero(_nonnegative(config["tolerance"]))
    maximum = _integer(config["maxVertices"], 1, _MAX_VERTICES)
    start_x, start_y = _zero(start_x0), _zero(start_y0)

    field = gradient_noise_2d_01({"seed": field_seed})
    configuration = {
        "field": field.serialize(), "start": [start_x, start_y], "heading": heading, "seed": seed,
        "attempts": tries, "stepDistance": distance, "fieldScale": scale,
        "fieldOffset": [offset_x, offset_y], "tolerance": tolerance, "maxVertices": maximum,
    }

    points = array("d", [start_x, start_y])
    directions = array("d", [])
    if tries == 0:
        return _Path(points, directions, 0, 0, 0, configuration)

    level = _query(field, start_x, start_y, scale, offset_x, offset_y, -1, "start_query")
    random = _Xoshiro(seed)
    x, y = start_x, start_y
    accepted, rejected = 0, 0

    for attempt in range(tries):
        low = (-_HALF_PI) * random.unit()
        high = _HALF_PI * random.unit()
        span = high - low
        delta = span * random.unit()
        turn = low + delta
        proposal = _arith(heading + turn, attempt, "proposal_heading")
        dx = _arith(distance * fdlibm_cos(proposal), attempt, "delta_x")
        next_x = _arith(x + dx, attempt, "position_x")
        dy = _arith(distance * fdlibm_sin(proposal), attempt, "delta_y")
        next_y = _arith(y + dy, attempt, "position_y")
        value = _query(field, next_x, next_y, scale, offset_x, offset_y, attempt, "candidate_query")
        drift_product = 0.2 * random.unit()
        drift = -0.1 + drift_product
        rejection_heading = _arith(heading + drift, attempt, "rejection_heading")
        difference = value - level
        band_error = abs(difference)
        if band_error < tolerance:
            if accepted + 1 == maximum:
                raise NoiseBandPathError("VERTEX_LIMIT_EXCEEDED", attempt, "append")
            next_x = _zero(next_x)
            next_y = _zero(next_y)
            canonical_proposal = _zero(proposal)
            points.append(next_x)
            points.append(next_y)
            directions.append(canonical_proposal)
            x, y, heading = next_x, next_y, canonical_proposal
            accepted += 1
        else:
            heading = _zero(rejection_heading)
            rejected += 1
    return _Path(points, directions, tries, accepted, rejected, configuration)

File: packages/javascript/src/noise-band-path.js
import { gradientNoise2D01 } from "./gradient-noise-2d-01.js";
import { fdlibmSin, fdlibmCos } from "./fdlibm-trig.js";

const HALF_PI = Math.PI / 2; // exactly 0x1.921fb54442d18p0 (halving is exact)
const UINT32_MAX = 4294967295;
const INT32_MAX = 2147483647;
const MAX_VERTICES = 1073741823;
const MAX_SAFE = 9007199254740991;
const KEYS = ["field", "start", "heading", "seed", "attempts", "stepDistance",
  "fieldScale", "fieldOffset", "tolerance", "maxVertices"];

/** Error with one of the stable path.noise-band-trace-2d contract codes. */
export class NoiseBandPathError extends Error {
  constructor(code, attemptIndex = null, stage = null) {
    super(code);
    this.name = "NoiseBandPathError";
    this.code = code;
    if (code === "TRACE_ARITHMETIC_INVALID" || code === "TRACE_QUERY_INVALID" || code === "VERTEX_LIMIT_EXCEEDED") {
      this.attemptIndex = attemptIndex;
      this.stage = stage;
    }
  }
}

function fail(code) { throw new NoiseBandPathError(code); }
function finite(value) { return typeof value === "number" && Number.isFinite(value); }
function zero(value) { return value === 0 ? 0 : value; }
function num(value) { if (!finite(value)) fail("INVALID_INPUT"); return value; }
function uint(value) {
  const n = num(value);
  if (n < 0 || n > UINT32_MAX || n !== Math.floor(n)) fail("INVALID_INPUT");
  return n;
}
function integer(value, lo, hi) {
  const n = num(value);
  if (n < lo || n > hi || n !== Math.floor(n)) fail("INVALID_INPUT");
  return n;
}
function nonnegative(value) {
  const n = num(value);
  if (n < 0) fail("INVALID_INPUT");
  return n;
}

function passiveRecord(value, keys) {
  if (value === null || typeof value !== "object" || Array.isArray(value)) fail("INVALID_INPUT");
  const prototype = Object.getPrototypeOf(value);
  if (prototype !== Object.prototype && prototype !== null) fail("INVALID_INPUT");
  if (Reflect.ownKeys(value).length !== keys.length) fail("INVALID_INPUT");
  for (const key of keys) {
    const descriptor = Object.getOwnPropertyDescriptor(value, key);
    if (descriptor === undefined || !("value" in descriptor)) fail("INVALID_INPUT");
  }
  return value;
}
function recordValue(record, key) { return Object.getOwnPropertyDescriptor(record, key).value; }
function passiveArray(value, length) {
  if (!Array.isArray(value) || Object.getPrototypeOf(value) !== Array.prototype) fail("INVALID_INPUT");
  if (length !== undefined && value.length !== length) fail("INVALID_INPUT");
  return value;
}
function arrayItem(array, index) {
  const descriptor = Object.getOwnPropertyDescriptor(array, String(index));
  if (descriptor === undefined || !("value" in descriptor)) fail("INVALID_INPUT");
  return descriptor.value;
}
function writableSlot(array, index) {
  const key = String(index);
  const descriptor = Object.getOwnPropertyDescriptor(array, key);
  if (descriptor !== undefined) return "value" in descriptor && descriptor.writable === true;
  if (!Object.isExtensible(array)) return false;
  for (let prototype = Object.getPrototypeOf(array); prototype !== null; prototype = Object.getPrototypeOf(prototype)) {
    if (Object.getOwnPropertyDescriptor(prototype, key) !== undefined) return false;
  }
  return true;
}
function pair(value) {
  const array = passiveArray(value, 2);
  return [zero(num(arrayItem(array, 0))), zero(num(arrayItem(array, 1)))];
}

function arith(value, attemptIndex, stage) {
  if (!Number.isFinite(value)) throw new NoiseBandPathError("TRACE_ARITHMETIC_INVALID", attemptIndex, stage);
  return value;
}
/** Q(x,y,prefix): field query with strict per-axis multiply-then-add domain checks. */
function query(field, x, y, scale, offsetX, offsetY, attemptIndex, prefix) {
  let qx = x * scale;
  if (!Number.isFinite(qx)) throw new NoiseBandPathError("TRACE_QUERY_INVALID", attemptIndex, prefix + "_x");
  qx = qx + offsetX;
  if (!Number.isFinite(qx) || qx < -MAX_SAFE || qx >= MAX_SAFE) {
    throw new NoiseBandPathError("TRACE_QUERY_INVALID", attemptIndex, prefix + "_x");
  }
  let qy = y * scale;
  if (!Number.isFinite(qy)) throw new NoiseBandPathError("TRACE_QUERY_INVALID", attemptIndex, prefix + "_y");
  qy = qy + offsetY;
  if (!Number.isFinite(qy) || qy < -MAX_SAFE || qy >= MAX_SAFE) {
    throw new NoiseBandPathError("TRACE_QUERY_INVALID", attemptIndex, prefix + "_y");
  }
  return field.sample(qx, qy);
}

// Private xoshiro128**1.1 stream with two SplitMix64 expansion words; same
// algorithm as design/operations/noise-band-path-contract.md's frozen RNG.
class Xoshiro {
  constructor(seed) {
    let x = BigInt(seed) + 0x9e3779b97f4a7c15n;
    const a = this.split(x);
    x += 0x9e3779b97f4a7c15n;
    const b = this.split(x);
    this.s0 = Number(a & 0xffffffffn) >>> 0;
    this.s1 = Number(a >> 32n) >>> 0;
    this.s2 = Number(b & 0xffffffffn) >>> 0;
    this.s3 = Number(b >> 32n) >>> 0;
  }
  split(x) {
    x = BigInt.asUintN(64, x);
    x = BigInt.asUintN(64, (x ^ (x >> 30n)) * 0xbf58476d1ce4e5b9n);
    x = BigInt.asUintN(64, (x ^ (x >> 27n)) * 0x94d049bb133111ebn);
    return BigInt.asUintN(64, x ^ (x >> 31n));
  }
  output() {
    const value = Math.imul(this.s1, 5) >>> 0;
    const result = Math.imul(((value << 7) | (value >>> 25)) >>> 0, 9) >>> 0;
    const t = (this.s1 << 9) >>> 0;
    this.s2 = (this.s2 ^ this.s0) >>> 0;
    this.s3 = (this.s3 ^ this.s1) >>> 0;
    this.s1 = (this.s1 ^ this.s2) >>> 0;
    this.s0 = (this.s0 ^ this.s3) >>> 0;
    this.s2 = (this.s2 ^ t) >>> 0;
    this.s3 = ((this.s3 << 11) | (this.s3 >>> 21)) >>> 0;
    return result;
  }
  unit() { return this.output() / 4294967296; }
}

function serializable(value) {
  if (Array.isArray(value)) return value.map(serializable);
  if (value !== null && typeof value === "object") {
    const out = {};
    for (const key of Object.keys(value)) out[key] = serializable(value[key]);
    return out;
  }
  return value;
}

class Path {
  #positions; #headings; #attempts; #accepted; #rejected; #config;
  constructor(positions, headings, attempts, accepted, rejected, config) {
    this.#positions = positions;
    this.#headings = headings;
    this.#attempts = attempts;
    this.#accepted = accepted;
    this.#rejected = rejected;
    this.#config = config;
    Object.freeze(this);
  }
  get size() { return this.#positions.length / 2; }
  get attempts() { return this.#attempts; }
  get accepted() { return this.#accepted; }
  get rejected() { return this.#rejected; }

  #checked(index, headings) {
    if (typeof index !== "number" || !Number.isFinite(index) || index < 0
        || index !== Math.floor(index) || index > MAX_SAFE) fail("INVALID_INDEX");
    if (index >= (headings ? this.#accepted : this.size)) fail("INDEX_OUT_OF_RANGE");
    return index;
  }
  pointAt(index) {
    const i = this.#checked(index, false);
    return [this.#positions[2 * i], this.#positions[2 * i + 1]];
  }
  pointInto(index, output, offset = 0) {
    const i = this.#checked(index, false);
    if (!Array.isArray(output) && !(output instanceof Float64Array)) fail("INVALID_OUTPUT");
    if (!Number.isInteger(offset) || offset < 0 || offset > output.length - 2) fail("INVALID_OUTPUT");
    if (!writableSlot(output, offset) || !writableSlot(output, offset + 1)) fail("INVALID_OUTPUT");
    output[offset] = this.#positions[2 * i];
    output[offset + 1] = this.#positions[2 * i + 1];
    return output;
  }
  headingAt(index) {
    const i = this.#checked(index, true);
    return this.#headings[i];
  }
  serialize() { return serializable(this.#config); }
  toValues() {
    const positions = [];
    for (let i = 0; i < this.size; i += 1) positions.push([this.#positions[2 * i], this.#positions[2 * i + 1]]);
    return {
      positions,
      headings: this.#headings.slice(),
      attempts: this.#attempts,
      accepted: this.#accepted,
      rejected: this.#rejected,
    };
  }
}

/**
 * Retain an attempt-bounded connected path whose accepted proposals remain within
 * a strict scalar band around the starting noise value. Implements
 * path.noise-band-trace-2d 0.1.0 independently of source code. Motivating sketch:
 * survey/out/2018/Generativos/venas; no artistic attempts/tolerance/scale range is
 * established, see the catalog contract for evidence. Uses a bit-exact ported
 * fdlibm5.3 sin/cos (verified against java.lang.StrictMath.sin/cos), not host Math.
 */
export function noiseBandPath2D(input) {
  const record = passiveRecord(input, KEYS);
  const fieldConfig = passiveRecord(recordValue(record, "field"), ["seed"]);
  const fieldSeed = uint(recordValue(fieldConfig, "seed"));
  const [startX0, startY0] = pair(recordValue(record, "start"));
  let heading = zero(num(recordValue(record, "heading")));
  const seed = uint(recordValue(record, "seed"));
  const tries = integer(recordValue(record, "attempts"), 0, INT32_MAX);
  const distance = zero(nonnegative(recordValue(record, "stepDistance")));
  const scale = zero(num(recordValue(record, "fieldScale")));
  const [offsetX, offsetY] = pair(recordValue(record, "fieldOffset"));
  const tolerance = zero(nonnegative(recordValue(record, "tolerance")));
  const maximum = integer(recordValue(record, "maxVertices"), 1, MAX_VERTICES);
  const startX = zero(startX0), startY = zero(startY0);

  const field = gradientNoise2D01({ seed: fieldSeed });
  const configuration = {
    field: field.serialize(), start: [startX, startY], heading, seed, attempts: tries,
    stepDistance: distance, fieldScale: scale, fieldOffset: [offsetX, offsetY],
    tolerance, maxVertices: maximum,
  };

  let points = [startX, startY];
  let directions = [];
  if (tries === 0) return new Path(points, directions, 0, 0, 0, configuration);

  const level = query(field, startX, startY, scale, offsetX, offsetY, -1, "start_query");
  const random = new Xoshiro(seed);
  let x = startX, y = startY;
  let accepted = 0, rejected = 0;

  for (let attempt = 0; attempt < tries; attempt += 1) {
    const low = (-HALF_PI) * random.unit();
    const high = HALF_PI * random.unit();
    const span = high - low;
    const delta = span * random.unit();
    const turn = low + delta;
    const proposal = arith(heading + turn, attempt, "proposal_heading");
    const dx = arith(distance * fdlibmCos(proposal), attempt, "delta_x");
    let nextX = arith(x + dx, attempt, "position_x");
    const dy = arith(distance * fdlibmSin(proposal), attempt, "delta_y");
    let nextY = arith(y + dy, attempt, "position_y");
    const value = query(field, nextX, nextY, scale, offsetX, offsetY, attempt, "candidate_query");
    const driftProduct = 0.2 * random.unit();
    const drift = -0.1 + driftProduct;
    const rejectionHeading = arith(heading + drift, attempt, "rejection_heading");
    const difference = value - level;
    const bandError = Math.abs(difference);
    if (bandError < tolerance) {
      if (accepted + 1 === maximum) throw new NoiseBandPathError("VERTEX_LIMIT_EXCEEDED", attempt, "append");
      nextX = zero(nextX);
      nextY = zero(nextY);
      const canonicalProposal = zero(proposal);
      points.push(nextX, nextY);
      directions.push(canonicalProposal);
      x = nextX; y = nextY; heading = canonicalProposal;
      accepted += 1;
    } else {
      heading = zero(rejectionHeading);
      rejected += 1;
    }
  }
  return new Path(points, directions, tries, accepted, rejected, configuration);
}
